pkg/etl: Add the embedded template library (etl)

jenswet opened the pull request:
### Contribution description

This PR contains the [ETL](https://www.etlcpp.com/) as package. It offers a similar API to the C++ stdlib without using the heap. Additionally it reimplements some parts of the stdlib like type-traits if there is no stdlib available.

I've used the ETL in embedded projects with RIOT for the containers and liked it.
Since I am currently implementing some C++ additions to RIOT I noticed that the AVR platform has no stdlib support (and therefore no type traits). Type traits are really convenient even without stdlib support. When using the etl package it's possible to take advantage of type traits during compile time without increasing ram or rom size.

### Testing procedure
The ETL is already heavily unit tested with several compilers. [[1](https://www.etlcpp.com/home.html)][[2](https://github.com/ETLCPP/etl/tree/master/test)]

I've just added one little test using some type traits, a vector and a string. This internally uses multiple parts of the ETL to verify that compilation works with multiple platforms. The test compares that the printed string is valid.

jenswet commented:
@maribu please also have a look. Thanks!

jenswet commented on pkg/etl/doc.txt:
```diff
@@ -0,0 +1,11 @@
+/**
+ * @defgroup pkg_etl Embedded Template Library
+ * @ingroup  pkg
```
Since this package is purely for C++ we could think about adding it to cpp group instead of pkg group. What do you think?

jenswet replied:
I think it's better to put it to C++, I changed it because no one has expressed a different opinion yet.

jenswet commented on pkg/etl/Kconfig:
```diff
@@ -0,0 +1,14 @@
+
+config PACKAGE_ETL
+    bool "Embedded Template Library"
+    depends on TEST_KCONFIG
```
Murdock error:
> /tmp/dwq.0.8167323357704165/4d8b5b6a5b794425a415c79057f65e0a/Makefile.base:80: *** Found C++ source, but "cpp" module is not used. Enable the MODULE_CPP in Kconfig.  Stop.
```suggestion
    depends on MODULE_CPP
    depends on TEST_KCONFIG
```
Do I need to add this here?

benpicco replied:
let's try it :wink:

jenswet replied:
Seems to be wrong.
```
[genconfig.py]:ERROR-=> The parameter PACKAGE_ETL could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: MODULE_CPP (=n)
```
Could anyone with Kconfig knowledge give a hint? Thanks! Maybe @MrKevinWeiss who had a good hint on this before?

miri64 replied:
Just do what `genconfig.py` is telling you: It complains that `MODULE_CPP` is set to `n`, so set it to `y` ;-).

miri64 replied:
(Kconfig does not pull in dependencies automatically, it will just warn you about it when you did not select the module).

benpicco commented on pkg/etl/config/etl_profile.h:
```diff
@@ -0,0 +1,39 @@
+*              This is configured by compiler defines the in Makefile.include
+* @author      Jens Wetterich <jens@wetterich-net.de>
+*/
+#if !defined(ETL_PROFILE_H) && !defined(DOXYGEN)
```
```suggestion
#ifndef ETL_PROFILE_H
```

Just use the standard header guard, doxygen will be fine.

jenswet replied:
The intention was to not need to document the defines. They are only for internal use in the ETL and not to be used anywhere in RIOT... 🤔

jenswet replied:
Changed it accordingly and documented the defines as `@internal`.

aabadie commented on pkg/etl/doc.txt:
```diff
@@ -0,0 +1,11 @@
+/**
+ * @defgroup cpp_etl Embedded Template Library
+ * @ingroup  cpp
```
Since it's a package it should at least be listed under the pkg group. You can also put it under cpp. The group name should be pkg_etl IMHO.

aabadie commented on tests/pkg_etl/Makefile.ci:
```diff
@@ -0,0 +1,2 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    nucleo-l011k4
```
```suggestion
    nucleo-l011k4 \
    #
```

jenswet replied:
Changed it. Could you explain what this is for?

aabadie replied:
Adding a new board after the nucleo-l011k4 will only change one line. That's minor but cleaner.

aabadie commented on tests/pkg_etl/app.config.test:
```diff
@@ -0,0 +1,2 @@
+CONFIG_MODULE_CPP=y
```
Better use `select MODULE_CPP` under the `config PACKAGE_ETL` definition than here.

```suggestion
```

jenswet replied:
Makes more sense, thanks!

jenswet commented:
@aabadie I've added a fixup commit that you can check the suggested changes. :)

Seems like Murdock fails with lots of Rust errors not related to my PR. I already did a rebase on the latest master. Is there something broken?

aabadie commented:
> Is there something broken?

Yes, we are discussing on how to fix it.

jenswet commented:
> > Is there something broken?
> 
> Yes, we are discussing on how to fix it.

Seems to be fixed. I squashed and rebased. :-)

jenswet commented on pkg/etl/Makefile.include:
```diff
@@ -0,0 +1,25 @@
+# So it's declared as a pseudo-module
+PSEUDOMODULES += etl
+
+# Activate the usage of libstdcpp types and features if available
+# This prevents unnecessary reimplementations
+ifneq (,$(filter libstdcpp,$(FEATURES_USED)))
+  CXXEXFLAGS += -DFEATURE_LIBSTDCPP
+endif
+
+# Some boards don't provide NAN or by default don't define limits like UINT8_MAX in CPP context
+# This activates the appropriate workarounds
+ifneq (,$(filter arch_avr8,$(FEATURES_USED)))
+  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
+  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
+else ifneq (,$(filter arch_riscv,$(FEATURES_USED)))
+  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
+  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
+endif
```
These settings are not activated when building with Kconfig.

Do I need to check for the Kconfig specific defines extra? Unfortunately, I haven't really understood how the Kconfig build works and interacts with the Makefiles yet.

```suggestion
# Activate the usage of libstdcpp types and features if available
# This prevents unnecessary reimplementations
ifneq (,$(filter libstdcpp,$(FEATURES_USED)))
  CXXEXFLAGS += -DFEATURE_LIBSTDCPP
# Kconfig alternatives
else ifndef HAS_LIBSTDCPP
  CXXEXFLAGS += -DFEATURE_LIBSTDCPP
endif

# Some boards don't provide NAN or by default don't define limits like UINT8_MAX in CPP context
# This activates the appropriate workarounds
ifneq (,$(filter arch_avr8,$(FEATURES_USED)))
  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
else ifneq (,$(filter arch_riscv,$(FEATURES_USED)))
  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
# Kconfig alternatives
else ifndef CPU_ARCH_AVR8
  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
else ifndef CPU_ARCH_RISCV
  CXXEXFLAGS += -DNO_CPP_NAN_SUPPORT
  CXXEXFLAGS += -D__STDC_LIMIT_MACROS
endif
```

jenswet replied:
Should be fixed by 74b7a26 hopefully.

jenswet commented on tests/pkg_etl/Makefile.ci:
```diff
@@ -0,0 +1,3 @@
+BOARD_INSUFFICIENT_MEMORY := \
+    nucleo-l011k4 \
```
These two have insufficient memory too.
```suggestion
    nucleo-l011k4 \
    stm32f030f4-demo \
    samd10-xmini \
```

jenswet reviewed (commented):
Could someone have a look on the Kconfig review comment?
Almost all boards (except the two with insufficient memory) failed because of some missing Kconfig setting in my Makefile probably.

maribu commented:
The issues with the compilation on AVR is due to the unmaintained and horribly outdated toolchain in the Debian / Ubuntu repositories. https://github.com/RIOT-OS/RIOT/pull/17479 is likely fixing those.

Interestingly, I cannot build this with picolibc (both with my local toolchain and `riot/riotbuild` docker image). This can be reproduced using:

```diff
diff --git a/tests/pkg_etl/Makefile b/tests/pkg_etl/Makefile
index 32e9011560..1da6f2d0b0 100644
--- a/tests/pkg_etl/Makefile
+++ b/tests/pkg_etl/Makefile
@@ -1,3 +1,6 @@
 include ../Makefile.tests_common
 USEPKG += etl
+
+FEATURES_REQUIRED += picolibc
+
 include $(RIOTBASE)/Makefile.include
```

(Or `FEATURES_REQUIRED=picolibc make`, but `FEATURES_REQUIRED` is not passed through to docker.)

jenswet commented:
> The issues with the compilation on AVR is due to the unmaintained and horribly outdated toolchain in the Debian / Ubuntu repositories. #17479 is likely fixing those.

@maribu I don't think there is an issue with the toolchain. I can build for AVR without problems locally and via `riot/riotbuild`. It seems to be related to Kconfig only:
```
[... Kconfig build errors removed for shorter text ....]
make[1]: *** [/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/Makefile.base:166: /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/application_tests_pkg_etl/main.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/tests/pkg_etl/../../Makefile.include:703: application_tests_pkg_etl.module] Error 2
make: Leaving directory '/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/tests/pkg_etl'
An error occurred while compiling using Kconfig
make: Entering directory '/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/tests/pkg_etl'
Building application "tests_pkg_etl" for "arduino-mega2560" with MCU "atmega2560".
rm -rf /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/pkg-build/etl

[INFO] cloning etl
git-cache: cloning from cache. tag=commitd3975ee36a826686f3a264a6e532818f9f79109e-773353
[INFO] updating etl /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/pkg/etl/.pkg-state.git-downloaded
echo d3975ee36a826686f3a264a6e532818f9f79109e  > /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/pkg/etl/.pkg-state.git-downloaded
[INFO] patch etl
sha1sum /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/tests/pkg_etl/tests/01-run.py /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/tests_pkg_etl.elf > /tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/test-input-hash.sha1
   text	   data	    bss	    dec	    hex	filename
   7872	    382	    965	   9219	   2403	/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/build/tests_pkg_etl.elf
make: Leaving directory '/tmp/dwq.0.449754928936169/53e3027bec133ebccd2ab3644fef27a9/tests/pkg_etl'
Hashes of binaries with and without Kconfig mismatch for tests/pkg_etl with arduino-mega2560
Please check that all used modules are modelled in Kconfig and enabled
```
Could you have a look at https://github.com/RIOT-OS/RIOT/pull/17477#discussion_r780647880?
> Interestingly, I cannot build this with picolibc (both with my local toolchain and `riot/riotbuild` docker image). This can be reproduced using:
> 
> ```diff
> diff --git a/tests/pkg_etl/Makefile b/tests/pkg_etl/Makefile
> index 32e9011560..1da6f2d0b0 100644
> --- a/tests/pkg_etl/Makefile
> +++ b/tests/pkg_etl/Makefile
> @@ -1,3 +1,6 @@
>  include ../Makefile.tests_common
>  USEPKG += etl
> +
> +FEATURES_REQUIRED += picolibc
> +
>  include $(RIOTBASE)/Makefile.include
> ```
> 
> (Or `FEATURES_REQUIRED=picolibc make`, but `FEATURES_REQUIRED` is not passed through to docker.)

I can't reproduce this.
I can build without issues for `BOARD=hifive1b` using `picolib` with `riot/riotbuild`.

jenswet commented:
@maribu @aabadie Murdock is green now. Except for the `6lowpan-clicker` board everything is fine, but this is related to the bug reported in #17573. Since the board is deprecated I think it's fine to keep it blacklisted.

Anything else you'd like to change? If not, I could squash and I'd be happy for an ACK. Thanks!

maribu commented on pkg/etl/config/etl_profile.h:
```diff
@@ -0,0 +1,53 @@
+*
+* @file
+* @brief       Common settings for etl
+* @details     This activates the usage of the STL if available and/or the usage of workarounds for missing NAN macros.
```
```suggestion
* @details     This activates the usage of the STL if available and/or the usage of
               workarounds for missing NAN macros.
```

The CI is very upset when lines are longer than 100 chars :/

maribu commented on tests/pkg_etl/main.cpp:
```diff
@@ -0,0 +1,61 @@
+    etl::to_string(teststr.size(), teststr, true);
+
+    /* Test string compare */
+    teststr += teststr == "Test: is_same: 1, is_unsigned: 0, vector: cdeab, size: 55" ? " -> valid" : " -> false";
```
Again, the CI is unhappy due to the 100 chars. Maybe?

```suggestion
    if (teststr == "Test: is_same: 1, is_unsigned: 0, vector: cdeab, size: 55" ? ") {
        teststr += -> valid";
    }
    else {
        teststr += " -> false";
    }
```

maribu reviewed (commented):
Looks good to me. The CI found two instances of more than 100 chars. Care to split those lines to get the CI happy? If so, please squash in right away.

I will run the provided test on a few platforms that I have at hands. If all is fine, I'd say this can go in.

jenswet commented:
@maribu Thanks for the feedback. I changed it accordingly. Additionally I've updated etl to the latest version and squashed.

Please let me know the result of your tests.

jenswet commented:
Murdock shows errors but they don't seem related to my changes and also failed in other recent PR builds...?

benpicco commented:
Yes the issue is known, but the cause is not.

maribu commented:
Test application works fine on a Nucleo-F767ZI and an Arduino Mega2560

maribu commented:
Thanks for the contribution!
